contracts/src/l1/TokenBridge.sol

```diff
@@ -197,103 +224,158 @@ contract TokenBridge is
+
+        if (
+            emissionPerSecond == 0 ||
+            totalSupply == 0 ||
+            lastUpdateTimestamp == block.timestamp ||
+            lastUpdateTimestamp >= distributionEnd
+        ) {
+            return index;
+        }
 
-        l1Token.safeTransferFrom(msg.sender, address(this), amount);
-        sendMessage(l1Token, msg.sender, l2Recipient, amount, block.number, rewardsIndex);
+        uint256 currentTimestamp = block.timestamp > distributionEnd
+            ? distributionEnd
+            : block.timestamp;
+        uint256 timeDelta = currentTimestamp.sub(lastUpdateTimestamp);
+        return
+            emissionPerSecond
+                .mul(timeDelta)
+                .mul(10**uint256(18))
+                .div(totalSupply)
+                .add(index); // 18- precision, should be loaded
     }
 
-    function depositUnderlying(
-        IStaticATokenLM l1Token,
+    function deposit(
+        address l1AToken,
         uint256 l2Recipient,
         uint256 amount,
-        uint16 refferalCode,
+        uint16 referralCode,
         bool fromAsset
-    )
-        onlyApprovedToken(l1Token)
-        onlyValidL2Address(l2Recipient)
-        external
-    {
-        uint256 rewardsIndex = l1Token.getCurrentRewardsIndex();
+    ) external onlyValidL2Address(l2Recipient) {
+        IERC20 underlyingAsset = aTokenData[l1AToken].underlyingAsset;
+        ILendingPool lendingPool = aTokenData[l1AToken].lendingPool;
+        require(
+            (underlyingAsset != IERC20(0x0)) &&
+                (lendingPool != ILendingPool(0x0)),
+            "This aToken has not been approved yet."
+        );
+
+        uint256 rewardsIndex = getCurrentRewardsIndex(l1AToken);
 
         if (fromAsset) {
-          l1Token.ASSET().safeTransferFrom(msg.sender, address(this), amount);
+            underlyingAsset.safeTransferFrom(msg.sender, address(this), amount);
+            lendingPool.deposit(
+                address(underlyingAsset),
+                amount,
+                address(this),
+                referralCode
+            );
         } else {
-          l1Token.ATOKEN().safeTransferFrom(msg.sender, address(this), amount);
+            IERC20(l1AToken).safeTransferFrom(
+                msg.sender,
+                address(this),
+                amount
+            );
         }
-
-        amount = l1Token.deposit(address(this), amount, refferalCode, fromAsset);
-        sendMessage(l1Token, msg.sender, l2Recipient, amount, block.number, rewardsIndex);
+        sendMessage(
+            l1AToken,
+            msg.sender,
+            l2Recipient,
+            _dynamicToStaticAmount(
+                amount,
+                address(underlyingAsset),
+                lendingPool
+            ),
+            block.number,
+            rewardsIndex
+        );
     }
 
-    function withdrawUnderlying(
-        IStaticATokenLM l1Token,
+    function withdraw(
```

Comment:
This method is missing a LogWithdraw, also we should consider returning the dynamic amount that has been withdrawn by this call

Reply:
the event is emitted in the function `consumeMessage` - which is misleading because we emit the event at the very beginning of the function